src/types/any.rs

```diff
@@ -146,36 +146,24 @@ impl PyAny {
     where
         O: ToPyObject,
     {
-        unsafe fn do_compare(
-            py: Python,
-            a: *mut ffi::PyObject,
-            b: *mut ffi::PyObject,
-        ) -> PyResult<Ordering> {
-            let result = ffi::PyObject_RichCompareBool(a, b, ffi::Py_EQ);
-            if result == 1 {
+        let py = self.py();
+        // Almost the same as ffi::PyObject_RichCompareBool, but this one doesn't try self == other.
+        // See https://github.com/PyO3/pyo3/issues/985 for more.
+        let do_compare = |other, op| unsafe {
+            PyObject::from_owned_ptr_or_err(py, ffi::PyObject_RichCompare(self.as_ptr(), other, op))
+                .and_then(|obj| obj.is_true(py))
+        };
+        other.with_borrowed_ptr(py, |other| {
+            if do_compare(other, ffi::Py_EQ)? {
                 return Ok(Ordering::Equal);
-            } else if result < 0 {
-                return Err(PyErr::fetch(py));
-            }
-            let result = ffi::PyObject_RichCompareBool(a, b, ffi::Py_LT);
-            if result == 1 {
+            } else if do_compare(other, ffi::Py_LT)? {
                 return Ok(Ordering::Less);
-            } else if result < 0 {
-                return Err(PyErr::fetch(py));
-            }
-            let result = ffi::PyObject_RichCompareBool(a, b, ffi::Py_GT);
-            if result == 1 {
+            } else if do_compare(other, ffi::Py_GT)? {
                 return Ok(Ordering::Greater);
-            } else if result < 0 {
-                return Err(PyErr::fetch(py));
             }
             Err(TypeError::py_err(
                 "PyAny::compare(): All comparisons returned false",
             ))
```

Comment:
I just realised we could change this to:

```suggestion
            } else {
                Err(TypeError::py_err(
                    "PyAny::compare(): All comparisons returned false",
                ))
            }
```

Then we could also remove the `return` keyword from all the other branches.

Reply:
:+1: